aws/src/main/java/org/apache/iceberg/aws/LakeFormationAwsClientFactory.java

```diff
@@ -0,0 +1,230 @@
+import org.apache.iceberg.relocated.com.google.common.base.Splitter;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.util.PropertyUtil;
+import software.amazon.awssdk.auth.credentials.AwsCredentials;
+import software.amazon.awssdk.auth.credentials.AwsCredentialsProvider;
+import software.amazon.awssdk.auth.credentials.AwsSessionCredentials;
+import software.amazon.awssdk.auth.credentials.StaticCredentialsProvider;
+import software.amazon.awssdk.http.urlconnection.UrlConnectionHttpClient;
+import software.amazon.awssdk.regions.PartitionMetadata;
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.services.dynamodb.DynamoDbClient;
+import software.amazon.awssdk.services.glue.GlueClient;
+import software.amazon.awssdk.services.glue.model.GetTableRequest;
+import software.amazon.awssdk.services.glue.model.GetTableResponse;
+import software.amazon.awssdk.services.kms.KmsClient;
+import software.amazon.awssdk.services.lakeformation.LakeFormationClient;
+import software.amazon.awssdk.services.lakeformation.model.GetTemporaryGlueTableCredentialsRequest;
+import software.amazon.awssdk.services.lakeformation.model.GetTemporaryGlueTableCredentialsResponse;
+import software.amazon.awssdk.services.s3.S3Client;
+import software.amazon.awssdk.services.sts.StsClient;
+import software.amazon.awssdk.services.sts.auth.StsAssumeRoleCredentialsProvider;
+import software.amazon.awssdk.services.sts.model.AssumeRoleRequest;
+import software.amazon.awssdk.services.sts.model.Tag;
+
+public class LakeFormationAwsClientFactory
+    implements AwsClientFactory {
+  private Map<String, String> properties;
+  private Region region;
+  private String s3Endpoint;
+  private String accessKeyId;
+  private String secretAccessKey;
+  private String sessionToken;
+
+  public LakeFormationAwsClientFactory() {
+  }
+
+  @Override
+  public void initialize(Map<String, String> catalogProperties) {
+    this.properties = Maps.newHashMap(catalogProperties);
+    this.region = region(properties);
+    this.s3Endpoint = properties.get(AwsProperties.S3FILEIO_ENDPOINT);
+    this.accessKeyId = properties.get(AwsProperties.S3FILEIO_ACCESS_KEY_ID);
+    this.secretAccessKey = properties.get(AwsProperties.S3FILEIO_SECRET_ACCESS_KEY);
+    this.sessionToken = properties.get(AwsProperties.S3FILEIO_SESSION_TOKEN);
+  }
+
+  @Override
+  public S3Client s3() {
+    return S3Client.builder()
+        .endpointOverride(URI.create(s3Endpoint))
+        .credentialsProvider(buildS3CredentialProvider())
+        .region(region)
+        .httpClientBuilder(UrlConnectionHttpClient.builder())
+        .build();
+  }
+
+  @Override
+  public GlueClient glue() {
+    return GlueClient.builder()
+        .credentialsProvider(getAwsCredentialsProvider())
+        .region(region)
+        .httpClientBuilder(UrlConnectionHttpClient.builder())
+        .build();
+  }
+
+  @Override
+  public KmsClient kms() {
```

Comment:
I think we also need to configure the KMS client? can you confirm?

Reply:
temporary credentials vended by LakeFormation have LakeFormation SLR as principal which is scopes down to S3 actions only